### PR TITLE
Add a page describing how we deploy lambdas

### PR DESCRIPTION
**Pull request by nodanaonlyzuul**

### CODE REVIEW

**Comment by nonword** [comment]
tldr: The requirement that I MUST install node-lambda locally creates a problem for `run`ing an app locally if it has aws-sdk dependencies.

I use `node-lambda run ...` to test code locally against a few sample `event.json`s as a last sanity check before deploying. Very strangely, `node-lambda run` does not accept this `--profile` flag. Without such a flag, `node-lambda` seems reliant on what's in the environment/`.env`. The issue I tried to describe yesterday was that there's a load order issue in `node-lambda` that initializes aws-sdk *before* it loads `.env`. (See https://github.com/motdotla/node-lambda/pull/369  ) Because aws-sdk is loaded before `.env` is loaded, it (sometimes?) holds onto ENV creds. In my case, this means when my app attempts to decrypt a string using kms, it's using the creds in my ENV - ignoring what's in `.env`. This issue is resolved if I run `node-lambda` globally - I think because the app obtains it's own instance of the aws-sdk credentials cache? 

So, without the ability to run `node-lambda run` against different accounts, my app can't decrypt kms strings encrypted using different accounts. If others are using kms (or other aws sdks) in their apps, I'd be interested to take a look at how you're able to run `./node_modules/.bin/node-lambda run` without it using the wrong aws creds. I would love to start running this locally, so I hope someone has pointers.

**Comment by nodanaonlyzuul** [comment]
Hey @nonword.

Good catch and a great explanation of the issue.
@rhernand3z calls `./node_modules/.bin/node-lambda run ...snip` [here](https://github.com/NYPL/nypl-hold-request-consumer/blob/master/package.json#L8) so I wonder if he's just never encountered this issue.

I wonder if your PR would be enough to satisfy what I wrote above...because
I think it would be ideal to have neither `AWS_ACCESS_KEY_ID` nor `AWS_SECRET_ACCESS_KEY` in `.env` file - so so the .env file can be committed to source. Seems like for that to happen the `run` command may need to support `--profile` as a first class citizen. (I'm digging through the code to see how the `deploy` command parses the `--profile` flag now.

Maybe after a chat with you, me, and @rhernand3z we can figure out if your PR is enough but since it hasn't been published to npm it seems like we'll have to loosen my original comment to a SHOULD or call out that it's a MUST for `deploy`...but until this feature is added for `run` we're stuck using global node-lambda.

Thanks for the explanation again.

**Comment by nonword** [comment]
My apps fail when invoking aws-sdk kms decrypt. I think that may be what's different about my app: my app uses the aws-sdk.

**Comment by nonword** [comment]
When you're stuck, I highly recommend chattin with @holingpoon . She proposed removing my `[default]` aws profile and wouldn't you know it that worked. It seems that the mere presence of that `default` caused my local instance of aws-sdk to initialize on invalid (sandbox) credentials even though `node-lambda`'s call to `dotenv` should have overridden them.

So I rescind the local node-lambda challenge to this PR. I think I should have no problem running node-lambda locally now. See also https://github.com/NYPL/engineering-general/issues/39

**Comment by nodanaonlyzuul** [comment]
So - it seems that aws-sdk glommed onto `[default]` in a very sticky way that the `dotenv`
couldn't overwrite.

**Comment by nonword** [comment]
Yeah, except I'd place the blame on aws-sdk's ["global configuration" handling](https://github.com/aws/aws-sdk-js/issues/1276) rather than `dotenv`. `dotenv` does exactly what it's supposed to, but if you call `dotenv` *after* initializing aws-sdk somewhere in your code (as node-lambda does), whatever happened to be in your environment (or `[default]` profile, as it turns out) is used in subsequent aws-sdk calls. Luckily, removing that `[default]` profile seems to prevent this early instantiation from holding onto anything and/or causes it to respect the new credentials that `dotenv` loads in.

We *will* still need a `--profile` flag on `node-lambda run` however if we're going with the model described above where acct creds are pulled from `~/.aws`. Until that's supported, I have no way to specify what profile to use when running `node-lambda run`.

**Comment by kfriedman** [comment]
All the AWS SDKs have a search order for finding credentials not explicitly provided to clients. And every SDK is a little different. I think we should ensure devs are aware of this behavior and the order in their particular SDK.

Node:
http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html
Java
http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html
PHP
http://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/credentials.html

**Comment by nodanaonlyzuul** [comment]
Sorry - I'm getting split between the conversation here & https://github.com/NYPL/engineering-general/issues/39. Any objections to me merging this in? I'm totally fine waiting.

**Comment by kfriedman** [comment]
I think merging is great. I just wanted to raise visibility about the AWS credential searching logic. Thanks for adding this!
